er/groovy/src/main/java/org/linqs/psl/examples/er/JaroWinklerSetSimilarity.java

```diff
@@ -0,0 +1,112 @@
+ * limitations under the License.
+ */
+package org.linqs.psl.examples.er;
+
+import org.linqs.psl.database.DatabaseQuery;
+import org.linqs.psl.database.ReadableDatabase;
+import org.linqs.psl.database.ResultList;
+import org.linqs.psl.model.term.Variable;
+import org.linqs.psl.model.function.ExternalFunction;
+import org.linqs.psl.model.predicate.StandardPredicate;
+import org.linqs.psl.model.term.Constant;
+import org.linqs.psl.model.term.ConstantType;
+import org.linqs.psl.model.term.UniqueIntID;
+import org.linqs.psl.model.atom.QueryAtom;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.TreeSet;
+
+import org.apache.commons.text.similarity.JaroWinklerDistance;
+
+public class JaroWinklerSetSimilarity implements ExternalFunction {
+	// similarity threshold (default=0.5)
+	private double simThresh;
+
+	public JaroWinklerSetSimilarity() {
+		this.simThresh = 0.5;
+	}
+
+	public JaroWinklerSetSimilarity(double simThresh) {
+		this.simThresh = simThresh;
+	}
+
+	@Override
+	public int getArity() {
+		return 2;
+	}
+
+	@Override
+	public ConstantType[] getArgumentTypes() {
+		return new ConstantType[] {ConstantType.UniqueIntID, ConstantType.UniqueIntID};
+	}
+
+	@Override
+	public double getValue(ReadableDatabase db, Constant... args) {
+		Set<String> authorNamesP1 = new HashSet<String>();
+		Set<String> authorNamesP2 = new HashSet<String>();
+		StandardPredicate authorOfPredicate = StandardPredicate.get("AuthorOf");
+		StandardPredicate authorNamePredicate = StandardPredicate.get("AuthorName");
+
+		DatabaseQuery query = new DatabaseQuery(new QueryAtom(authorOfPredicate, new Variable("A"), args[0]));
+		ResultList results = db.executeQuery(query);
+		for (int i = 0; i < results.size(); i++) {
+			query = new DatabaseQuery(new QueryAtom(authorNamePredicate, results.get(i)[0], new Variable("A")));
+			ResultList resultsTwo = db.executeQuery(query);
+			for (int j = 0; j < resultsTwo.size(); j++) {
+				authorNamesP1.add(String.valueOf(resultsTwo.get(j)[0]));
+			}
+		}
+
+		query = new DatabaseQuery(new QueryAtom(authorOfPredicate, new Variable("A"), args[1]));
+		results = db.executeQuery(query);
+		for (int i = 0; i < results.size(); i++) {
+			query = new DatabaseQuery(new QueryAtom(authorNamePredicate, results.get(i)[0], new Variable("A")));
+			ResultList resultsTwo = db.executeQuery(query);
+			for (int j = 0; j < resultsTwo.size(); j++) {
+				authorNamesP2.add(String.valueOf(resultsTwo.get(j)[0]));
+			}
+		}
+
+		if (authorNamesP2.size() < authorNamesP1.size()){
+			Set<String> authorNamesTemp = authorNamesP1;
+			authorNamesP1 = authorNamesP2;
+			authorNamesP2 = authorNamesTemp;
+		}
+
+		double total = 0.0;
+		double max = 0.0;
+		double sim = 0.0;
+		JaroWinklerDistance jaroW = new JaroWinklerDistance();
+
+		for (String s1 : authorNamesP1) {
+			String remove = "";
+			for (String s2 : authorNamesP2) {
+				sim = jaroW.apply(s1, s2);
+				if (sim > max){
```

Comment:
!!
This is bugged.
`max` never gets reset, so one round of comparisons will leak into the next.

Set `max = 0.0` at the same time as remove.
Set remove = null instead of empty string.
This if should be `remove == null || sim > max` (you want to catch the first iteration).